rfcs/text/0000-analytics-instrumentation-plan.md

```diff
@@ -0,0 +1,146 @@
+
+### Configuration
+
+1. We configure a [tracking ID](https://support.google.com/analytics/answer/1008080?hl=en) in the GA console for each of
+   the deployment environments
+    1. `dev` - for local development and tests on the dev deployment
+    2. `integration` - for integration tests
+    3. `staging` - for the staging environment
+    4. `prod` - for the production deployment and real usage in the browser
+
+Dividing up metrics collection by deployment environment allows us to do the following.
+
+* test that metrics are flowing for earlier deployment stages before they hit production
+* separate out real usage metrics from that generated by developers and tests
+
+### Event schema
+
+Sending custom analytics events to GA is possible from the client-side [via
+gtags.js](https://developers.google.com/analytics/devguides/collection/gtagjs/events) or from the server-side [via
+Measurement Protocol](https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide). To keep events
+consistent, issuers of custom events should follow the following schema.
+
+* `category` - the name of the service issuing the event, e.g. `data-storage-service`, `data-portal`, or `dcp-cli`
+* `action` - is a human-readable string representing the function performed, e.g. `download-bundle` or `upload-file`
+* `label` - is the identifier of the entity acted on, e.g, bundle UUID or `<organ>-<method>-<data-contributor-last-name>`*
+
+### Frontend instrumentation
+
+For frontend components, we can use the standard GA instrumentation. Steps are included below.
+
+1. Front-end developers create trackers using the standard JavaScript tracking snippet    ([see
+   documentation](https://developers.google.com/analytics/devguides/collection/gtagjs/)).
+2. Tracking IDs should be chosen based on its deployment environment.
+3. At this time, we suggest not customizing the tracking behavior in any way. We expect the default settings will be
+   sufficient for our needs, and we can iterate if they are not after we’ve captured early user data.
+
+### Metrics Exporter
+
+The metrics exporter is an application for exporting request/response information from server-side logs to Google
+Measurement Protocol API in batch. Its components are as follows.
+
+* A [Kinesis Data Firehose Stream to
+  s3](https://docs.aws.amazon.com/firehose/latest/dev/create-destination.html#create-destination-s3) (KDFS)
+* A lambda that listens to new s3 events generated by Kinesis Data Firehose, loads the request information and submits
+  it   to the Google Measurement Protocol API   [in
+  batch](https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide#batch)
+
+### Data storage service instrumentation
+
+The Data Storage Service (DSS) should fire an event whenever data is requested and pushed to users. Steps are included
+below.
+
+* Upon each `GET /v1/bundles/:bundle_uuid` and `GET /v1/files/:file_uuid` request, a JSON log line is emitted to
```

Comment:
Some of our CLI commands use stdout (`hca dss get-bundle` for instance). Does that interfere with this? Will the analytics output be part of my get-bundle results?

Reply:
This concerns server-side logging.

We discussed and cleared this up with @kozbo during the 2019-03-22 Architecture Team Meeting.